independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/DirectoryPathTree.java

```diff
@@ -55,8 +73,30 @@ public void walk(PathVisitor visitor) {
         PathTreeVisit.walk(dir, dir, pathFilter, getMultiReleaseMapping(), visitor);
     }
 
+    private String ensureRelative(String path) {
+        if (path == null) {
+            return path;
+        }
+        if (isAbsolutePath(path)) {
+            final Path absolutePath = dir.getFileSystem().getPath(path).normalize().toAbsolutePath();
+            if (absolutePath.startsWith(dir)) {
+                return dir.relativize(absolutePath).toString();
+            }
```

Comment:
I would not do this.

This behavior is actually the current one, and something we need to avoid in order to fix #23574. The user was setting a configuration property to something like `/home/yrodiere/workspace/myapp/src/main/resources/myfile.txt`. We used `DirectoryPathTree` to check that the resource exist and is a file (not a directory), and the check passed because `DirectoryPathTree` was accepting absolute paths and interpreting them as relative to the FS root. However, later on when we tried to retrieve the resource from the classpath, it didn't work anymore, because `/home/yrodiere/workspace/myapp/src/main/resources/myfile.txt` obviously isn't a valid path to a resource in the classpath.

Also, if you think about it, this behavior allows for Quarkus configurations that may or may not work depending on where on the filesystem the application code is checked out... which seems very bad. For example, if I set some Quarkus config property to `/home/yrodiere/workspace/myapp/src/main/resources/myfile.txt`, and the string eventually ends up in this method, it will be relativized automatically if I run this code on my machine, but not on yours where the application is located in `/home/aloubyansky/workspace/myapp`.

What you are doing is accepting absolute paths as long as they target a file within the directory tree.
What we should be doing, IMO, is what I did in #23672: whenever we get an absolute path, we interpret it as relative to `dir`, i.e. relative to the root of the classpath. In short absolute paths would be absolute paths *in the classpath*, not in the filesystem. Which seems to be what `DirectoryPathTree` is about anyway?

So the code would become:

```suggestion
            if (absolutePath.startsWith(dir)) {
                return absolutePath.relativize(absolutePath.getRoot()).toString();
            }
```

Reply:
Wouldn't it better to simply throw an `IllegalArgumentException` in case `dir` is not equal to `absolutePath.getRoot()`?
I find it counter intuitive otherwise.

Reply:
Actually, it wouldn't help in your case.

Reply:
It's just there could be two different valid paths: one relative to the actual root of the passed in path and one relative to the `dir`. In which case, it's kind of confusing, especially on Windows, where the absolute path includes the disk part.
Perhaps, we should throw an IllegalArgumentException if we detect the disk part?

Reply:
I think in my case (#23574), passing an path such as `/my-resource.sql` could make sense: the user might just mean "absolute path to a resource", so "absolute" as in "relative to the root of the classpath". Of course in #23574 that's not what they did; and there's no way to differentiate between a user mistakenly passing an absolute (unix) filesystem path and a user passing an absolute resource path. So I'd say absolute unix paths should probably be allowed and handled like in my suggestion?

Passing a windows absolute path with a disk part, on the other hand... yeah, that doesn't make sense and would deserve an exception.